source/extensions/load_balancing_policies/common/load_balancer_impl.cc

```diff
@@ -637,18 +640,43 @@ absl::FixedArray<ZoneAwareLoadBalancerBase::LocalityPercentages>
 ZoneAwareLoadBalancerBase::calculateLocalityPercentages(
     const HostsPerLocality& local_hosts_per_locality,
     const HostsPerLocality& upstream_hosts_per_locality) {
-  uint64_t total_local_hosts = 0;
-  std::map<envoy::config::core::v3::Locality, uint64_t, LocalityLess> local_counts;
+  std::map<envoy::config::core::v3::Locality, uint64_t, LocalityLess> local_weights;
+  std::map<envoy::config::core::v3::Locality, uint64_t, LocalityLess> upstream_weights;
+  uint64_t total_local_weight = 0;
   for (const auto& locality_hosts : local_hosts_per_locality.get()) {
-    total_local_hosts += locality_hosts.size();
+    uint64_t locality_weight = 0;
+    // If locality_basis_ is set to HOST_WEIGHT, we use the host's weight to calculate the locality
+    // percentage. Otherwise, we count the number of hosts in the locality.
+    if (locality_basis_ == LocalityLbConfig::ZoneAwareLbConfig::HOST_WEIGHT) {
+      for (const auto& host : locality_hosts) {
+        locality_weight += host->weight();
+      }
+    } else {
+      locality_weight = locality_hosts.size();
+    }
+    total_local_weight += locality_weight;
     // If there is no entry in the map for a given locality, it is assumed to have 0 hosts.
     if (!locality_hosts.empty()) {
-      local_counts.insert(std::make_pair(locality_hosts[0]->locality(), locality_hosts.size()));
+      local_weights.insert(std::make_pair(locality_hosts[0]->locality(), locality_weight));
     }
   }
-  uint64_t total_upstream_hosts = 0;
+  uint64_t total_upstream_weight = 0;
   for (const auto& locality_hosts : upstream_hosts_per_locality.get()) {
-    total_upstream_hosts += locality_hosts.size();
+    uint64_t locality_weight = 0;
+    // If locality_basis_ is set to HOST_WEIGHT, we use the host's weight to calculate the locality
+    // percentage. Otherwise, we count the number of hosts in the locality.
+    if (locality_basis_ == LocalityLbConfig::ZoneAwareLbConfig::HOST_WEIGHT) {
```

Comment:
See the switch comment above. It would be good to try and avoid duplicating this block if it's straightforward.